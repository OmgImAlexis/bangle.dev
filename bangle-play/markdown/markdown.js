import { extensions } from '../app/editor/extensions';
import * as markdown from 'bangle-plugins/markdown/index';
import { getSchema } from 'bangle-play/app/editor/utils';
import { defaultMarkdownItTokenizer } from 'bangle-plugins/markdown/index';
import { emojiMarkdownItPlugin } from 'bangle-plugins/emoji/emoji-markdown-it-plugin';

// TODO this is ugly read up about the static extensions though piece
const exts = extensions();
const schema = getSchema(exts);
const parser = markdown.markdownParser(
  extensions(),
  schema,
  defaultMarkdownItTokenizer.use(emojiMarkdownItPlugin),
);
const serializer = markdown.markdownSerializer(extensions());

export const markdownParser = (markdownStr) => parser.parse(markdownStr);
export const markdownSerializer = (doc) => serializer.serialize(doc);
