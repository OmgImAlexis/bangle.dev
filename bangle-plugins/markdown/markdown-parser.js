import markdownit from 'markdown-it';
import { MarkdownParser } from 'prosemirror-markdown';

export function markdownParser(schema) {
  return new MarkdownParser(
    schema,
    markdownit(),
    // .use(todoListPlugin),
    tokens,
  );
}

const tokens = {
  blockquote: { block: 'blockquote' },
  paragraph: { block: 'paragraph' },
  list_item: { block: 'list_item' },
  bullet_list: { block: 'bullet_list' },
  ordered_list: {
    block: 'ordered_list',
    getAttrs: (tok) => ({ order: +tok.attrGet('start') || 1 }),
  },
  heading: {
    block: 'heading',
    getAttrs: (tok) => ({ level: tok.tag.slice(1) }),
  },
  code_block: { block: 'code_block', noCloseToken: true },
  fence: {
    block: 'code_block',
    getAttrs: (tok) => ({ language: tok.info || '' }),
    noCloseToken: true,
  },
  hr: { node: 'horizontal_rule' },
  image: {
    node: 'image',
    getAttrs: (tok) => ({
      src: tok.attrGet('src'),
      title: tok.attrGet('title') || null,
      alt: (tok.children[0] && tok.children[0].content) || null,
    }),
  },
  hardbreak: { node: 'hard_break' },

  s: { mark: 'strike' },
  em: { mark: 'italic' },
  strong: { mark: 'bold' },
  link: {
    mark: 'link',
    getAttrs: (tok) => ({
      href: tok.attrGet('href'),
      title: tok.attrGet('title') || null,
    }),
  },
  code_inline: { mark: 'code', noCloseToken: true },
};

// function todoListPlugin(md) {
//   function isInline(token) {
//     return token.type === 'inline';
//   }
//   function isParagraph(token) {
//     return token.type === 'paragraph_open';
//   }
//   function isListItem(token) {
//     return token.type === 'list_item_open';
//   }
//   function startsWithTodoMarkdown(token) {
//     // leading whitespace in a list item is already trimmed off by markdown-it
//     return (
//       token.content.indexOf('[ ] ') === 0 ||
//       token.content.indexOf('[x] ') === 0 ||
//       token.content.indexOf('[X] ') === 0
//     );
//   }
//   function isTodoItem(tokens, index) {
//     return (
//       isInline(tokens[index]) &&
//       isParagraph(tokens[index - 1]) &&
//       isListItem(tokens[index - 2]) &&
//       startsWithTodoMarkdown(tokens[index])
//     );
//   }

//   function todoify(token, TokenConstructor) {
//     token.children.unshift(makeCheckbox(token, TokenConstructor));
//     token.children[1].content = token.children[1].content.slice(3);
//     token.content = token.content.slice(3);

//     if (useLabelWrapper) {
//       if (useLabelAfter) {
//         token.children.pop();

//         // Use large random number as id property of the checkbox.
//         var id =
//           'task-item-' + Math.ceil(Math.random() * (10000 * 1000) - 1000);
//         token.children[0].content =
//           token.children[0].content.slice(0, -1) + ' id="' + id + '">';
//         token.children.push(afterLabel(token.content, id, TokenConstructor));
//       } else {
//         token.children.unshift(beginLabel(TokenConstructor));
//         token.children.push(endLabel(TokenConstructor));
//       }
//     }
//   }
//   md.core.ruler.after('inline', 'todo_list', (state) => {
//     const tokens = state.tokens;

//     // var tokens = state.tokens;
//     for (var i = 2; i < tokens.length; i++) {
//       if (isTodoItem(tokens, i)) {
//         todoify(tokens[i], state.Token);
//         attrSet(
//           tokens[i - 2],
//           'class',
//           'task-list-item' + (!disableCheckboxes ? ' enabled' : ''),
//         );
//         attrSet(
//           tokens[parentToken(tokens, i - 2)],
//           'class',
//           'contains-task-list',
//         );
//       }
//     }
//   });
// }
