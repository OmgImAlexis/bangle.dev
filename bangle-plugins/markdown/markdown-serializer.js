import { objectFilter, objectMapValues } from 'bangle-core/utils/js-utils';
import { MarkdownSerializer } from 'prosemirror-markdown';

// A markdown serializer which uses a node/mark schema's
// toMarkdown property to generate a markdown string
export const markdownSerializer = (
  schema,
  nodeOverrides = {}, // Object<string, function>
  markdownOverrides = {}, // Object<string, function>
) => {
  return new MarkdownSerializer(
    proxyNodesToMarkdown(nodeOverrides, schema),
    proxyMarksToMarkdown(markdownOverrides, schema),
    schema,
  );
};

function proxyNodesToMarkdown(obj, schema) {
  const handler = {
    get(target, propKey, receiver) {
      const override = Reflect.get(target, propKey, receiver);
      if (override) {
        return override;
      }

      if (schema.nodes[propKey]) {
        if (schema.nodes[propKey].spec?.toMarkdown) {
          return schema.nodes[propKey].spec?.toMarkdown;
        }
      }

      return undefined;
    },
  };
  return new Proxy(obj, handler);
}

function proxyMarksToMarkdown(obj, schema) {
  const handler = {
    get(target, propKey, receiver) {
      const override = Reflect.get(target, propKey, receiver);
      if (override) {
        return override();
      }

      if (schema.marks[propKey]) {
        if (schema.marks[propKey].spec?.toMarkdown) {
          return schema.marks[propKey].spec?.toMarkdown();
        }
      }

      return undefined;
    },
  };
  return new Proxy(obj, handler);
}

export function serializeAtomNodeToMdLink(name, attrs) {
  const data = objectFilter(attrs, (val, key) => {
    return key.startsWith('data-');
  });

  const string = new URLSearchParams(
    objectMapValues(data, (val) => {
      // convert it to string for predictability when parsing different types
      return JSON.stringify(val);
    }),
  );

  return `[$${name}](bangle://${string})`;
}
