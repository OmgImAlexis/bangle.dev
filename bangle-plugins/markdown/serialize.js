import { MarkdownSerializer } from 'prosemirror-markdown';

// A markdown serializer which uses a node/mark schema's
// toMarkdown property to generate a markdown string
export const markdownSerializer = (
  schema,
  nodeOverrides = {}, // Object<string, function>
  markdownOverrides = {}, // Object<string, function>
) => {
  return new MarkdownSerializer(
    proxyNodesToMarkdown(nodeOverrides, schema),
    proxyMarksToMarkdown(markdownOverrides, schema),
    schema,
  );
};

function proxyNodesToMarkdown(obj, schema) {
  let handler = {
    get(target, propKey, receiver) {
      const override = Reflect.get(target, propKey, receiver);
      if (override) {
        return override;
      }

      if (schema.nodes[propKey]) {
        if (schema.nodes[propKey].spec?.toMarkdown) {
          return schema.nodes[propKey].spec?.toMarkdown;
        }
      }

      return undefined;
    },
  };
  return new Proxy(obj, handler);
}

function proxyMarksToMarkdown(obj, schema) {
  let handler = {
    get(target, propKey, receiver) {
      const override = Reflect.get(target, propKey, receiver);
      if (override) {
        return override();
      }

      if (schema.marks[propKey]) {
        if (schema.marks[propKey].spec?.toMarkdown) {
          return schema.marks[propKey].spec?.toMarkdown();
        }
      }

      return undefined;
    },
  };
  return new Proxy(obj, handler);
}
