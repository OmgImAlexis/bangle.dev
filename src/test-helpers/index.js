'use strict';

import prettier from 'prettier';
import {
  EditorContextProvider,
  EditorContext,
  EditorOnReadyContext,
} from 'utils/bangle-utils/helper-react/editor-context';
import { ReactEditor } from 'utils/bangle-utils/helper-react/react-editor';
import React from 'react';
import { render, waitForElement } from '@testing-library/react';
import pmTestBuilder from 'prosemirror-test-builder';
import browser from 'utils/bangle-utils/utils/browser';
import { Text, Paragraph, Doc } from 'utils/bangle-utils/nodes/index';
import {
  EditorState,
  Selection,
  TextSelection,
  NodeSelection,
} from 'prosemirror-state';
import { ExtensionManager } from 'utils/bangle-utils/utils/extension-manager';
import { Schema } from 'prosemirror-model';

export async function renderReactEditor(options = {}, testId = 'test-editor') {
  const _options = {
    id: 'test-editor',
    editorProps: {
      attributes: { class: 'bangle-editor content' },
    },
    extensions: [...(options.extensions || [])],
    doc: options.testDoc,
    selection: options.testDoc && selectionFor(options.testDoc),
    ...options,
  };

  let _editor;

  const result = render(
    <EditorContextProvider>
      <ReactEditor options={_options} />
      <EditorContext.Consumer>
        {(context) => {
          if (context.editor && !_editor) {
            _editor = context.editor;
          }
          return !context.editor ? null : <span data-testid={testId} />;
        }}
      </EditorContext.Consumer>
    </EditorContextProvider>,
  );
  const update = (doc) =>
    _editor.view.updateState(
      EditorState.create({
        doc: doc,
        selection: selectionFor(doc),
        schema: _editor.state.schema,
        plugins: _editor.state.plugins,
        storedMarks: _editor.state.storedMarks,
      }),
    );

  await result.findByTestId('test-editor');
  return {
    ...result,
    editor: _editor,
    builders: pmTestBuilder.builders(_editor.schema, {
      doc: { nodeType: 'doc' },
      p: { nodeType: 'paragraph' },
      text: { nodeType: 'text' },
      li: { nodeType: 'list_item' },
      ul: { nodeType: 'bullet_list' },
      ol: { nodeType: 'ordered_list' },
    }),
    update,
  };
}

function selectionFor(doc) {
  let a = doc.tag.a;
  if (a != null) {
    let $a = doc.resolve(a);
    if ($a.parent.inlineContent)
      return new TextSelection(
        $a,
        doc.tag.b != null ? doc.resolve(doc.tag.b) : undefined,
      );
    else return new NodeSelection($a);
  }
  return Selection.atStart(doc);
}

export const jestExpect = {
  toEqualDoc,
};

expect.extend(jestExpect);

function toEqualDoc(received, expected) {
  if (received.type.name !== 'doc') {
    throw new Error('Received is not a valid PM doc. Please wrap with doc()');
  }
  if (expected.type.name !== 'doc') {
    throw new Error('expected is not a valid PM doc. Please wrap with doc()');
  }

  const frmt = (doc) =>
    prettier.format(doc.toString(), {
      semi: false,
      parser: 'babel',
      printWidth: 40,
      singleQuote: true,
    });
  const options = {
    comment: 'PM Doc .eq equality',
    isNot: this.isNot,
    promise: this.promise,
  };

  const pass = received.eq(expected);
  let message;
  if (pass) {
    message = () =>
      this.utils.matcherHint('toEqualDoc', undefined, undefined, options) +
      '\n\n' +
      `Expected: not ${this.utils.printExpected(expected.toJSON())}\n` +
      `Received: ${this.utils.printReceived(received.toJSON())}`;
  } else {
    const diffString = this.utils.diff(frmt(expected), frmt(received), {
      expand: this.expand,
    });
    message = () =>
      this.utils.matcherHint('toEqualDoc', undefined, undefined, options) +
      '\n\n' +
      (diffString && diffString.includes('- Expect')
        ? `Difference:\n\n${diffString}`
        : `Expected: ${this.utils.printExpected(expected.toJSON())}\n` +
          `Received: ${this.utils.printReceived(received.toJSON())}`);
  }
  return {
    pass,
    message,
  };
}

export function sendKeyToPm(editorView, keys) {
  const keyCodes = {
    'Enter': 13,
    'Backspace': 8,
    'Tab': 9,
    'Shift': 16,
    'Ctrl': 17,
    'Alt': 18,
    'Pause': 19,
    'CapsLock': 20,
    'Esc': 27,
    'Space': 32,
    'PageUp': 63276,
    'PageDown': 63277,
    'End': 63275,
    'Home': 63273,
    'Left': 63234,
    'Up': 63232,
    'Right': 63235,
    'Down': 63233,
    'PrintScrn': 44,
    'Insert': 63302,
    'Delete': 46,
    ';': 186,
    '=': 187,
    'Mod': 93,
    '*': 106,
    '-': 189,
    '.': 190,
    '/': 191,
    ',': 188,
    '`': 192,
    '[': 219,
    '\\': 220,
    ']': 221,
    "'": 222,
  };

  const event = new CustomEvent('keydown', {
    bubbles: true,
    cancelable: true,
  });
  event.DOM_KEY_LOCATION_LEFT = 1;
  event.DOM_KEY_LOCATION_RIGHT = 2;

  let parts = keys.split(/-(?!'?$)/);

  // set location property of event if Left or Right version of key specified
  let location = 0;
  const locationKeyRegex = /^(Left|Right)(Ctrl|Alt|Mod|Shift|Cmd)$/;
  parts = parts.map((part) => {
    if (part.search(locationKeyRegex) === -1) {
      return part;
    }
    const [, pLocation, pKey] = part.match(locationKeyRegex);
    location =
      pLocation === 'Left'
        ? event.DOM_KEY_LOCATION_LEFT
        : event.DOM_KEY_LOCATION_RIGHT;
    return pKey;
  });

  const modKey = parts.indexOf('Mod') !== -1;
  const cmdKey = parts.indexOf('Cmd') !== -1;
  const ctrlKey = parts.indexOf('Ctrl') !== -1;
  const shiftKey = parts.indexOf('Shift') !== -1;
  const altKey = parts.indexOf('Alt') !== -1;
  const key = parts[parts.length - 1];

  // all of the browsers are using the same keyCode for alphabetical keys
  // and it's the uppercased character code in real world
  const code = keyCodes[key] ? keyCodes[key] : key.toUpperCase().charCodeAt(0);

  event.key = key.replace(/Space/g, ' ');
  event.shiftKey = shiftKey;
  event.altKey = altKey;
  event.ctrlKey = ctrlKey || (!browser.mac && modKey);
  event.metaKey = cmdKey || (browser.mac && modKey);
  event.keyCode = code;
  event.which = code;
  event.view = window;
  event.location = location;

  editorView.dispatchEvent(event);
}

export function insertText(view, text, from, _to) {
  let pos = typeof from === 'number' ? from : view.state.selection.from;

  text.split('').forEach((character, index) => {
    if (
      !view.someProp('handleTextInput', (f) =>
        f(view, pos + index, pos + index, character),
      )
    ) {
      view.dispatch(
        view.state.tr.insertText(character, pos + index, pos + index),
      );
    }
  });
}

export function sleep(t = 20) {
  return new Promise((res) => setTimeout(res, t));
}

export function builders(extensions = []) {
  const schema = getSchema(extensions);
  return {
    schema,
    ...pmTestBuilder.builders(schema, {
      doc: { nodeType: 'doc' },
      p: { nodeType: 'paragraph' },
      text: { nodeType: 'text' },
      li: { nodeType: 'list_item' },
      ul: { nodeType: 'bullet_list' },
      ol: { nodeType: 'ordered_list' },
    }),
  };
}

export function getSchema(extensions = []) {
  const baseExtns = [new Doc(), new Text(), new Paragraph()];
  const manager = new ExtensionManager([...baseExtns, ...extensions], {});

  const schema = new Schema({
    topNode: 'doc',
    marks: manager.marks,
    nodes: manager.nodes,
  });

  return schema;
}
