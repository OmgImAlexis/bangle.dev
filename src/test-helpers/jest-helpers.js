import { Node } from 'prosemirror-model';

export const jestExpect = {
  toEqualDocAndSelection,
  toEqualDocument(actual, expected) {
    return toEqualDocument(
      this.equals,
      this.utils,
      this.expand,
    )(actual, expected);
  },
};

expect.extend(jestExpect);

function toEqualDocAndSelection(actual, expected) {
  const { doc: actualDoc, selection: actualSelection } = actual;
  const docComparison = toEqualDocument(
    this.equals,
    this.utils,
    this.expand,
  )(actualDoc, expected);
  if (!docComparison.pass) {
    return docComparison;
  }

  expected =
    typeof expected === 'function' && actualDoc.type && actualDoc.type.schema
      ? expected(actualDoc.type.schema)
      : expected;

  const fail = {
    pass: false,
    actual,
    expected,
    name: 'toEqualDocumentAndSelection',
    message: () => 'Expected specified selections to match in both values.',
  };

  if (expected.refs) {
    const refConditions = {
      '<': (position, selection) => position === selection.$from.pos,
      '>': (position, selection) => position === selection.$to.pos,
      '<>': (position, selection) =>
        position === selection.$from.pos && position == selection.$to.pos,
      '<node>': (position, selection) =>
        selection instanceof NodeSelection && position === selection.$from.pos,
      // The | denotes the gap cursor's side, based on the node on the side of the |.
      '<|gap>': (position, selection) =>
        // Using literal values from constructor as unable to import type from editor-core
        // Some tests use mock packages which will conflict with jestFrameworkSetup.js
        selection.constructor.name === 'GapCursorSelection' &&
        selection.side === 'right' &&
        position === selection.$from.pos,
      '<gap|>': (position, selection) =>
        selection.constructor.name === 'GapCursorSelection' &&
        selection.side === 'left' &&
        position === selection.$from.pos,
    };

    if (
      !Object.keys(refConditions).every((key) => {
        if (key in expected.refs) {
          return refConditions[key](expected.refs[key], actualSelection);
        }
        return true;
      })
    ) {
      return fail;
    }
  }

  return docComparison;
}

function toEqualDocument(equals, utils, expand) {
  return (actual, expected) => {
    // Because schema is created dynamically, expected value is a function (schema) => PMNode;
    // That's why this magic is necessary. It simplifies writing assertions, so
    // instead of expect(doc).toEqualDocument(doc(p())(schema)) we can just do:
    // expect(doc).toEqualDocument(doc(p())).
    //
    // Also it fixes issues that happens sometimes when actual schema and expected schema
    // are different objects, making this case impossible by always using actual schema to create expected node.
    expected =
      typeof expected === 'function' && actual.type && actual.type.schema
        ? expected(actual.type.schema)
        : expected;

    if (!(expected instanceof Node) || !(actual instanceof Node)) {
      return {
        pass: false,
        actual,
        expected,
        name: 'toEqualDocument',
        message: () =>
          'Expected both values to be instance of prosemirror-model Node.',
      };
    }

    if (expected.type.schema !== actual.type.schema) {
      return {
        pass: false,
        actual,
        expected,
        name: 'toEqualDocument',
        message: () => 'Expected both values to be using the same schema.',
      };
    }

    const pass = equals(actual.toJSON(), expected.toJSON());
    const message = pass
      ? () =>
          `${utils.matcherHint('.not.toEqualDocument')}\n\n` +
          `Expected JSON value of document to not equal:\n  ${utils.printExpected(
            expected,
          )}\n` +
          `Actual JSON:\n  ${utils.printReceived(actual)}`
      : () => {
          const diffString = utils.diff(expected, actual, {
            expand: expand,
          });
          return (
            `${utils.matcherHint('.toEqualDocument')}\n\n` +
            `Expected JSON value of document to equal:\n${utils.printExpected(
              expected,
            )}\n` +
            `Actual JSON:\n  ${utils.printReceived(actual)}` +
            `${diffString ? `\n\nDifference:\n\n${diffString}` : ''}`
          );
        };

    return {
      pass,
      actual,
      expected,
      message,
      name: 'toEqualDocument',
    };
  };
}
