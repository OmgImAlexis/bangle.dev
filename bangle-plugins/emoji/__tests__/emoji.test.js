/**
 * @jest-environment jsdom
 */
/** @jsx psx */
import { fireEvent, wait } from '@testing-library/react';
import { psx, sendKeyToPm, renderTestEditor } from 'bangle-core/test-helpers/';
import {
  OrderedList,
  BulletList,
  ListItem,
  Heading,
  HardBreak,
  TodoList,
  TodoItem,
} from 'bangle-core/nodes';
import { markdownSerializer } from 'bangle-plugins/markdown/markdown-serializer';
import EmojiExtension from '../index';

const extensions = [
  new BulletList(),
  new ListItem(),
  new OrderedList(),
  new TodoList(),
  new TodoItem(),
  new HardBreak(),
  new Heading(),
  new EmojiExtension(),
];
const testEditor = renderTestEditor({ extensions });

test('Rendering works', async () => {
  Date.now = jest.fn(() => 0);
  const { container, editor } = await testEditor(
    <doc>
      <para>
        foo[]bar
        <emoji data-emojikind=":horse:" />
      </para>
    </doc>,
  );

  expect(editor.state).toEqualDocAndSelection(
    <doc>
      <para>
        foo[]bar
        <emoji data-emojikind=":horse:" />
      </para>
    </doc>,
  );
  expect(container.querySelector(`[data-type="emoji"]`)).toMatchSnapshot();
});

describe('markdown', () => {
  let schemaPromise;
  const serialize = async (doc) => {
    const content = doc(await schemaPromise);
    return markdownSerializer(await schemaPromise).serialize(content);
  };
  beforeAll(async () => {
    schemaPromise = renderTestEditor({ extensions })().then((r) => r.schema);
  });

  test('markdown serialization', async () => {
    const md = await serialize(
      <doc>
        <para>
          hello world
          <emoji data-emojikind=":horse:" />
        </para>
      </doc>,
    );
    expect(md).toMatchInlineSnapshot(
      `"hello world[$emoji](bangle://data-emojikind=%22%3Ahorse%3A%22&data-type=%22emoji%22)"`,
    );
  });
});
