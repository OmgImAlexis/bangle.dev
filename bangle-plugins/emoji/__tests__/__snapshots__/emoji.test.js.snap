// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering works 1`] = `
<span
  class="emoji-NodeView-Wrap"
  data-emojikind=":horse:"
  data-type="emoji"
  draggable="true"
  style="display: inline-block;"
>
  <span
    contenteditable="false"
  >
    🐰
  </span>
</span>
`;
